File: TP2/src/phrases.py
# Fonction pour compter le nombre d'occurrences d'une phrase dans une ligne
def compter_occurrences(ligne, phrase):
    return ligne.count(phrase)

# Demander à l'utilisateur le nom du fichier
nom_fichier = input("Entrez le nom du fichier : ")

# Demander à l'utilisateur la phrase à rechercher
phrase = input("Entrez la phrase que vous souhaitez rechercher : ")

# Ouvrir le fichier en mode lecture
try:
    with open(nom_fichier, 'r') as fichier:
        lignes = fichier.readlines()
        
        # Parcourir chaque ligne du fichier
        for num_ligne, ligne in enumerate(lignes, start=1):
            occurrences = compter_occurrences(ligne, phrase)
            
            # Si la phrase est présente, afficher le numéro de la ligne et le nombre d'occurrences
            if occurrences > 0:
                print(f"Ligne {num_ligne}, {occurrences} fois")
except FileNotFoundError:
    print(f"Erreur : Impossible d'ouvrir le fichier {nom_fichier}")

File: TP2/src/phrases.c
#include <stdio.h>
#include <string.h>

// Fonction pour compter le nombre d'occurrences d'une phrase dans une ligne donnée
int compter_occurrences(char *ligne, char *phrase) {
    int count = 0;
    char *pos = ligne;
    
    // Recherche de la phrase dans la ligne
    while ((pos = strstr(pos, phrase)) != NULL) {
        count++;
        pos += strlen(phrase); // Avancer le pointeur pour éviter les répétitions sur la même occurrence
    }
    
    return count;
}

int main() {
    char nom_fichier[100], phrase[100], ligne[256];
    FILE *fichier;
    int num_ligne = 0, occurrences;

    // Demande du nom du fichier
    printf("Entrez le nom du fichier : ");
    scanf("%s", nom_fichier);

    // Demande de la phrase à rechercher
    printf("Entrez la phrase que vous souhaitez rechercher : ");
    scanf(" %[^\n]s", phrase);  // %[^\n] permet de lire une chaîne avec des espaces

    // Ouverture du fichier en mode lecture
    fichier = fopen(nom_fichier, "r");
    if (fichier == NULL) {
        printf("Erreur : Impossible d'ouvrir le fichier %s\n", nom_fichier);
        return 1;
    }

    // Lecture du fichier ligne par ligne
    while (fgets(ligne, sizeof(ligne), fichier) != NULL) {
        num_ligne++;
        // Compter le nombre d'occurrences de la phrase dans la ligne
        occurrences = compter_occurrences(ligne, phrase);

        // Si la phrase est présente, afficher le numéro de la ligne et le nombre d'occurrences
        if (occurrences > 0) {
            printf("Ligne %d, %d fois\n", num_ligne, occurrences);
        }
    }

    // Fermeture du fichier
    fclose(fichier);

    return 0;
}
